### PR TITLE
Fix mishandling of revision "zero"

### LINKED ISSUE #2758: Bug: Invalid checkout commands generated by "Subversion checkout file"
Clicking the "Subversion checkout file" in admin mode generates a list of repositories and revisions to check out that satisfy the requirements for marking (such as deadlines); however, some students appear in the file as requiring revision 0 to be checked out (an invalid revision for Subversion). My guess is the student's account was created after the assignment was due or that this error occurs for all students that have not submitted anything to MarkUs or their first submission is late.

This error can cause problems if error codes from the checkout file are being tested (such as in a Makefile).

OS: MacOS 10.12 beta (build 16A313a)
Browser: Safari 10.0 (12602.1.50.0.8)
MarkUs version: 1.5.0.DEV
Ruby version: 2.2.5
Rails version: 4.2.4

### PR DESCRIPTION
**Pull request by adisandro**
* svn has a fake revision zero to represent the repo at initialization
time, checkout commands that use it are just wrong, the collected
revision handles it correctly
 * git has always an inital commit with an empty file, checkout commands
are correct, the collected revision shows it incorrectly

Fixes #2758.

### CODE REVIEW

**Comment by adisandro** [comment]
@david-yz-liu I can't fix this efficiently in git unless I change some assumptions, i.e. I can't make an empty submission not appear in the checkout commands without spending time walking commits.

I propose to formalize for all repo types what svn is already doing in a sneaky way, having a bogus revision "zero" for collected submissions with no files. We could either use 0 as revision identifier and timestamp, or allow nil in the database for those fields (probably more effort).

**Comment by david-yz-liu** [comment]
@adisandro hmm this is interesting. Two comments:

1. Assuming the "collected revision identifier" for git is the SHA of the first commit, is it not possible to replace the `== 0` check with a `== <SHA of first commit>` check? I.e., through the Rugged API determine the SHA of the first commit? (Analogous to the [command line](https://github.com/git-tips/tips#retrieve-the-commit-hash-of-the-initial-revision)) This seems like the easiest check to make.

2. If this is not possible, I think using `nil` is preferable to use `0` for all revision types. I'm not convinced this is that much more work than the existing approach.

**Comment by adisandro** [comment]
@david-yz-liu I'll tell you more in person tomorrow:
1. Yes, but you have to open the repo, create a reverse walker, and walk. Not so lightweight, not terribly expensive, but one order of magnitude slower than a check if submission.revision_identifier.nil? or == 0
2. nil is certainly clearer, but the assumption everywhere is that it is not allowed, so no checks are made.

**Comment by adisandro** [comment]
@david-yz-liu you can finally review this.
I found issues with how SVN gets revisions for a certain path (it's missing some related to deletions or empty dirs), but I think those problems existed before my changes. I may spend some more time on those.

File: app/jobs/autotest_run_job.rb [hunk 1]
```diff
@@ -7,7 +7,11 @@ class AutotestRunJob < ApplicationJob
   # Verify that MarkUs has student files to run the test.
   # Note: this does not guarantee all required files are presented.
   # Instead, it checks if there is at least one source file is successfully exported.
-  def repo_files_available?(assignment, repo_dir)
+  def repo_files_available?(assignment, submission, repo_dir)
```
[the remark this answers is not in the document]
Comment (adisandro) [reply app/jobs/autotest_run_job.rb]: ```submission``` may be nil here, for a test ran by a student

File: app/jobs/autotest_run_job.rb [hunk 2]
```diff
@@ -36,8 +40,8 @@ def perform(host_with_port, test_scripts, user_api_key, server_api_key, grouping
 
     # create empty test results for no submission files
     repo_dir = File.join(AutomatedTestsClientHelper::STUDENTS_DIR, group.repo_name)
-    unless repo_files_available?(assignment, repo_dir)
-      submission = submission_id.nil? ? nil : Submission.find(submission_id)
+    submission = if submission_id.nil? then nil else Submission.find(submission_id) end
```
[the remark this answers is not in the document]
Comment (adisandro) [reply app/jobs/autotest_run_job.rb]: Ok, I just don't like it when the condition itself ends with a ?

File: app/controllers/submissions_controller.rb [hunk 3]
```diff
@@ -633,27 +633,26 @@ def check_collect_status
...
-          "#{@grouping.group.group_name}_r#{@revision.revision_identifier}.zip"
+      begin
+        if revision_identifier.nil?
+          @revision = repo.get_latest_revision
+        else
+          @revision = repo.get_revision(revision_identifier)
+        end
+      rescue Repository::RevisionDoesNotExist
+        render text: t('student.submission.no_revision_available')
+        return true
```
[the remark this answers is not in the document]
Comment (adisandro) [reply app/controllers/submissions_controller.rb]: Not really, no.
I'm sorry, I just copied and pasted the error lines from ```download_subdirectories``` and forgot it there.

**Pull request merged**